dalm/pipelines/reading_comprehension_pipeline.py

```diff
@@ -0,0 +1,382 @@
+        # generate regex based reading comprehension dataset
+        regex_rc_gen = RegexBasedReadingComprehension(general_spm, domain_spm)
+
+        # NOTE: this is a simple check to see if the dataset is already generated
+        in_memory_dataset.extend(
+            [{"messages": rc_text} for _, _, rc_text in regex_rc_gen.dataset_generator(input, csv_column)]
+        )
+
+    # NOTE: this operation is time consuming and very expensive
+    # Attention has been paid to try to save intermediate steps in case of failure
+    # so that the generation can be resumed from the last checkpoint
+    if comprehension_type in [SynthMode.LLM, SynthMode.BOTH] and llm_kwargs:
+        if generation_state_file:
+            if os.path.exists(generation_state_file):
+                with open(generation_state_file, "rb") as f:
+                    generation_state = pickle.load(f)
+            else:
+                generation_state = {"processed_files": [], "total_files": 0, "files_missed": 0}
+                pickle.dump(generation_state, open(generation_state_file, "wb"))
+
+        if not os.path.exists(llm_kwargs.dataset_output_path):
+            os.makedirs(llm_kwargs.dataset_output_path)
+
+        llm_rc_dataset_generator = generate_synthetic_dataset(
+            model_name=llm_kwargs.model_name,
+            input_directory_or_file=input,
+            processed_files=generation_state["processed_files"],
+            chunk=llm_kwargs.chunk or False,
+            context_length=llm_kwargs.context_length or 0,
+            csv_column=csv_column,
+        )
+
+        # generate llm based reading comprehension dataset
+        for index, filename, context, gen_text in llm_rc_dataset_generator:
+            qanda = question_and_answer_extractor(gen_text, context)
+            if qanda:
+                output_file = f"{filename}_{index}.json"
+                with open(os.path.join(llm_kwargs.dataset_output_path, output_file), "w") as o:
+                    json.dump(qanda, o)
+            else:
+                logger.warning(
+                    (
+                        f"No question and answer pairs found for {filename} " f"chunk: {index}"
+                        if llm_kwargs.chunk
+                        else ""
+                    )
+                )
+                generation_state["files_missed"] += 1
+            generation_state["processed_files"].append(filename)
+            generation_state["total_files"] += 1
+            pickle.dump(generation_state, open(generation_state_file, "wb"))
+
+        logger.info(" Statistics ")
+        success_files_count = generation_state["total_files"] - generation_state["files_missed"]
+        logger.info(f"Total number of successfully extracted q&a: {success_files_count}")
+        logger.info(f"Total files missed: {generation_state['files_missed']} out of {generation_state['total_files']}")
+        logger.info(f"Total files processed: {generation_state['total_files']}")
```

Comment:
These stats are really helpful!  The term "files" is a bit misleading, but its clear from the output that can also represent rows.

If the dataset is empty however, we should throw an exception.

```
if not dataset:
    raise Exception("Failed to generate dataset")
```

This can happen when you have a small dataset and each chunk fails to generate a question/answer pair.

Reply:
I believe it will fail on it's own

```python
>>> datasets.Dataset.from_list([])
Dataset({
    features: [],
    num_rows: 0
})
>>> a = datasets.Dataset.from_list([])
>>> a.save_to_disk('./my_dataset')
Saving the dataset (0/1 shards): 0 examples [00:00, ? examples/s]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/xxx/Projects/trial/lib/python3.11/site-packages/datasets/arrow_dataset.py", line 1530, in save_to_disk
    for job_id, done, content in Dataset._save_to_disk_single(**kwargs):
  File "/home/xxx/Projects/trial/lib/python3.11/site-packages/datasets/arrow_dataset.py", line 1571, in _save_to_disk_single
    num_examples, num_bytes = writer.finalize()
                              ^^^^^^^^^^^^^^^^^
  File "/home/xxx/Projects/trial/lib/python3.11/site-packages/datasets/arrow_writer.py", line 599, in finalize
    raise SchemaInferenceError("Please pass `features` or at least one example when writing data")
datasets.arrow_writer.SchemaInferenceError: Please pass `features` or at least one example when writing data

```

but the message won't be clear so your recommendation stands

Reply:
Yeah I hit that error testing with a small dataset and it was a pretty cryptic failure message.  The idea is just to make it a lot clearer when that happens.